sklearn/feature_extraction/text.py

```diff
@@ -242,6 +242,8 @@ def build_analyzer(self):
     def _validate_vocabulary(self):
         vocabulary = self.vocabulary
         if vocabulary is not None:
+            if isinstance(vocabulary, set):
+                self.vocabulary = list(vocabulary)
```

Comment:
The problem is that ``fit`` is not supposed to change any of the initialization parameters.
The problem with the sorting is that if ``fit`` is called after pickling, I think this will not work, as the call to ``list`` will create lists with different ordering.